### PR TITLE
Set the user agent header in the Java client

### PR DESCRIPTION
**Pull request by mikaelm12**
Issue:#12295 
(Now based on master)

### CODE REVIEW

**Comment by mikaelm12** [comment]
From previous PR. 

A few things.

If it's for 3.1 and there's time, we'll do it in 3.0 :). This is motivated by Azure SignalR wanting this info, so it seems OK. If we can land this change this week, I don't see a concern with taking this in 3.0. It depends on what we can figure out from the below discussion though :). Let's get it right and then figure out what vehicle it will fit in.

Let's make the user agent values consistent across as many clients as we can. Right now, the .NET client's user agent is `Microsoft.AspNetCore.Http.Connections.Client/[version]`. Yuck. The layering makes sense, but I don't like it. My vote would be to just compromise the layering a bit here and call it "Microsoft SignalR". It's just a string, nothing actually couples the layers by doing this. A huge majority of people using the lower layer will be using "SignalR" as we know it.

My proposed format for the User-Agent string is something like this:

```
Microsoft SignalR/[Version] ([Detailed Version]; [Operating System]; [Runtime]; [Runtime Version])
```

Where:
* `[Version]` is the major/minor of the SignalR client itself.
  * For .NET client, we can just use the major and minor components of the Assembly version.
  * For JS we have a `VERSION` constant which is the full version. We can parse it, or add a new `SHORT_VERSION` constant or something with just the major/minor component.
  * For Java I don't know, we'll have to find out.
* `[Detailed Version]` is the really detailed version (including full build number). In .NET we store this in the `AssemblyInformationalVersionAttribute`. TBD for Java/JS
* `[Operating System]` is an OS type, just based on [`RuntimeInformation.IsOSPlatform`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.runtimeinformation.isosplatform?view=netframework-4.8) in .NET, `process.platform` in JS.
	* `Windows NT` for Windows
	* `macOS` for macOS
	* `Linux` for Linux
* `[Runtime]` is a simple name defining the kind of runtime the client is on (`.NET`, `Java`, `NodeJS` would be the three options we have right now)
* `[Runtime Version]` is a version of the runtime (.NET, Java, NodeJS, etc.):
	* .NET: [`RuntimeInformation.FrameworkDescription`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.runtimeinformation.frameworkdescription?view=netframework-4.8) is probably best
	* Java: Probably the [`java.version` system property](https://docs.oracle.com/javase/tutorial/essential/environment/sysprop.html), maybe with `java.vendor` prepended?
	* NodeJS: `process.versions.node`

Examples:

* .NET Core: `Microsoft SignalR/3.0 (3.0.0-preview9.12345.1; Windows NT; .NET; .NET Core 4.6.27317.03)` (the version number FrameworkDescription spits out is a bit odd but we can back-trace it to the "marketing" version)
* Java: `Microsoft SignalR/3.0 (3.0.0-preview9.12345.1; macOS; Java; 1.5.0)`
* NodeJS: `Microsoft SignalR/3.0 (3.0.0-preview9.12345.1; Linux; NodeJS; 11.11.0)`

So there are two work items here I think:

1. Figure out how to properly get the version of the Java client from **within the Java client itself*. We may actually have to just generate a Java class with it as a constant, if there's no built-in way (like our attributes in .NET).

1. Update all the user agents to have this data.

And of course, a user-specified `User-Agent` value should always override.

Thoughts?

File: src/SignalR/clients/java/signalr/build.gradle [hunk 1]
```diff
@@ -108,3 +108,27 @@ task generatePOM {
...
+        def versionFile = file("$buildDir/../src/main/java/com/microsoft/signalr/Version.java")
+        versionFile.parentFile.mkdirs()
+        versionFile.text =
+                """
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+package com.microsoft.signalr;
+
+public class Version {
```
Comment (anurse) [inline src/SignalR/clients/java/signalr/build.gradle]: Let's make this package-private, so that only code in the `com.microsoft.signalr` can see it i.e.:

```suggestion
class Version {
```

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/UserAgentHelper.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        return version;
+    }
+
+    static String findOSName(String operatingSystem) {
+        operatingSystem = operatingSystem.toLowerCase();
+        if (operatingSystem.indexOf("win") >= 0) {
+            return "Windows NT";
+        } else if (operatingSystem.contains("mac")) {
+            return "Mac";
+        } else if (operatingSystem.contains("nix") || operatingSystem.contains("nux") || operatingSystem.contains("aix")) {
```
Comment (anurse) [inline src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/UserAgentHelper.java]: 😆 I love it.

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/UserAgentHelper.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        }
+        return version;
+    }
+
+    static String findOSName(String operatingSystem) {
+        operatingSystem = operatingSystem.toLowerCase();
+        if (operatingSystem.indexOf("win") >= 0) {
+            return "Windows NT";
+        } else if (operatingSystem.contains("mac")) {
+            return "Mac";
```
Comment (anurse) [inline src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/UserAgentHelper.java]:
```suggestion
            return "macOS";
```

We should either use `Macintosh` (which is what browsers use), or `macOS` (the "Product Name" and what Apple is using everywhere). I prefer the latter. `Mac`, `Mac OS X` and `OSX` are all old news ;).

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/UserAgentHelper.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        if (detailedVersion.contains("-")) {
+            version = detailedVersion.substring(0, detailedVersion.indexOf('-'));
+        } else {
+            version = detailedVersion;
+        }
+        return version;
+    }
+
+    static String findOSName(String operatingSystem) {
+        operatingSystem = operatingSystem.toLowerCase();
```
Comment (anurse) [inline src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/UserAgentHelper.java]: We should see what this does on Android. If possible, it would be awesome to put in `Android` as the OS name and the actual Android version for the OS version.

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/Version.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+package com.microsoft.signalr;
+
+public class Version {
```
Comment (anurse) [inline src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/Version.java]:
```suggestion
class Version {
```

File: src/SignalR/clients/java/signalr/src/test/java/com/microsoft/signalr/HubConnectionTest.java [hunk 6]
```diff
@@ -2185,6 +2185,53 @@ public void connectionSendsPingsRegularly() throws InterruptedException {
...
+
+        hubConnection.start().timeout(1, TimeUnit.SECONDS).blockingAwait();
+        assertEquals(HubConnectionState.CONNECTED, hubConnection.getConnectionState());
+        hubConnection.stop();
+
+        assertTrue(header.get().startsWith("Microsoft SignalR/"));
+    }
+
+    @Test
+    public void userAgentHeaderCanBeOverwritten() {
```
Comment (anurse) [inline src/SignalR/clients/java/signalr/src/test/java/com/microsoft/signalr/HubConnectionTest.java]: Let's have another test verifying it can be "cleared" by setting to an empty string.

File: src/SignalR/clients/java/signalr/src/test/java/com/microsoft/signalr/UserAgentTest.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+import java.util.stream.Stream;
+
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+public class UserAgentTest {
+
+    private static Stream<Arguments> OperatingSystems() {
```
Comment (anurse) [inline src/SignalR/clients/java/signalr/src/test/java/com/microsoft/signalr/UserAgentTest.java]: Missing a test that uses this?

File: src/SignalR/clients/java/signalr/src/test/java/com/microsoft/signalr/UserAgentTest.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.util.stream.Stream;
+
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+public class UserAgentTest {
```
Comment (anurse) [inline src/SignalR/clients/java/signalr/src/test/java/com/microsoft/signalr/UserAgentTest.java]: Let's also add some tests that validate the Java Vendor and Version. We'll need to read those in the test so we know what to validate against. The test should also just strip off and ignore the version info (so we don't have to keep updating it).

File: src/SignalR/clients/java/signalr/src/test/java/com/microsoft/signalr/UserAgentTest.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+                Arguments.of("Macintosh", "Mac"),
+                Arguments.of("Linux", "Linux"),
+                Arguments.of("unix", "Linux"),
+                Arguments.of("", ""),
+                Arguments.of("1234", ""));
+    }
+
+    private static Stream<Arguments> Versions() {
+        return Stream.of(
+                Arguments.of("1.0.0", "1.0.0"),
```
Comment (anurse) [inline src/SignalR/clients/java/signalr/src/test/java/com/microsoft/signalr/UserAgentTest.java]: The short version should be just two segments: `1.0` or `3.1`. The patch version would be in the detailed version. The idea is that you can group on the short version and see all the clients with the same feature set.

File: src/SignalR/clients/java/signalr/src/test/java/com/microsoft/signalr/UserAgentTest.java (hunk 7 above)
Comment (mikaelm12) [reply src/SignalR/clients/java/signalr/src/test/java/com/microsoft/signalr/UserAgentTest.java]: Must have cut instead of copy :/

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/Version.java (hunk 5 above)
Comment (mikaelm12) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/Version.java]: It has to be changed in the gradle task that generates this file otherwise it will just get overwritten.

Comment (anurse) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/Version.java]: Yep, but we should change it in both.

Comment (mikaelm12) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/Version.java]: Yeah, changing in the other will change it here 😄

Comment (anurse) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/Version.java]: Oh, the generated file is checked in? We probably shouldn't do that unless absolutely necessary.

Comment (anurse) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/Version.java]: I had assumed that what was happening was the task was rewriting this file, but if it's generating it, we probably don't need to have it in Git at all

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/UserAgentHelper.java (hunk 4 above)
Comment (mikaelm12) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/UserAgentHelper.java]: Just did some searching and it seems like there are some other ways (java.vendor for example) to check if you're running on Android but I'm curious to see what all the different values are. I'll spin up Android studio and check it out.

Comment (halter73) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/UserAgentHelper.java]: How ugly would it be if we just included the os.name property directly in the user agent?

Comment (mikaelm12) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/UserAgentHelper.java]: AndroidDebugValues

Comment (mikaelm12) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/UserAgentHelper.java]: The vendor seems like the best way to determine if we're running on Android. FWIW Stackoverflow agrees.

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/UserAgentHelper.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+        if (operatingSystem.indexOf("win") >= 0) {
+            return "Windows NT";
+        } else if (operatingSystem.contains("mac")) {
+            return "macOS";
+        } else if (operatingSystem.contains("nix") || operatingSystem.contains("nux") || operatingSystem.contains("aix")) {
+            return "Linux";
+        }
+
+        return "";
```
Comment (BrennanConroy) [inline src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/UserAgentHelper.java]: Should this be empty?

**Review by BrennanConroy — COMMENTED** [review]
Can we somehow have a test that tests the actual user-agent string?

For example, I don't see where the parentheses are being added in `createUserAgentString` and a test would make that obvious

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/UserAgentHelper.java (hunk 4 above)
Comment (halter73) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/UserAgentHelper.java]: I agree that we should include both os and vendor info in the UA. The question I have is why do we need the findOSName at all instead of appending os.name directly?

In the Android case, it seems that os.name is already "Linux" so findOSName does nothing making this a moot point. But what if someone really is running on AIX or something weird, but it just shows "Linux" in the user agent? I'd rather not lose information unless os.name is sometimes extremely ugly and/or verbose.

Comment (mikaelm12) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/UserAgentHelper.java]: I don't really feel strongly either way. I understand the desire to standardize in case it's verbose like you mentioned but we could also just use the value directly. I'm fine with either, just don't want to get hung up on this.

Comment (mikaelm12) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/UserAgentHelper.java]: An argument for standardizing could be it makes it easier to test

Comment (halter73) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/UserAgentHelper.java]: My vote is for including the unaltered os.name. The tests can reference os.name and just make sure it matches and is in the right spot. I agree it's not a big deal though.

**Review by halter73 — APPROVED** [review]

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/UserAgentHelper.java (hunk 10 above)
Comment (mikaelm12) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/UserAgentHelper.java]: Just going to use the full `os.name` description now.

**Comment by mikaelm12** [comment]
@aspnet-hello triage https://dev.azure.com/dnceng/public/_build/results?buildId=318930

**Comment by aspnet-hello** [comment]
This comment was made automatically. If there is a problem contact aspnetcore-build@microsoft.com.

I've triaged the above build. I've created/commented on the following issue(s)
https://github.com/aspnet/AspNetCore-Internal/issues/2730

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/UserAgentHelper.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        return USER_AGENT;
+    }
+
+    public static String createUserAgentString() {
+        StringBuilder agentBuilder = new StringBuilder("Microsoft SignalR/");
+
+        // Parsing version numbers
+        String detailedVersion = Version.getDetailedVersion();
+        agentBuilder.append(getVersion(detailedVersion));
+        agentBuilder.append("; (");
```
Comment (BrennanConroy) [inline src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/UserAgentHelper.java]: I think we don't want this `;`
